### PR TITLE
Download gems with threads.

### PR DESCRIPTION
**Pull request by indirect**
This PR supercedes #649, implementing threaded downloads for gems. (Implemented while pairing with @duckinator).

---

(Following was added by @duckinator:)

TODO:

5. [ ] Tests!
4. [x] Use `locked_puts` in [`ThreadedDownloadReporter#fetch`](https://github.com/duckinator/rubygems/blob/db4a30ac3b5d901900e0f6777fc0b20bf5822fb0/lib/rubygems/user_interaction.rb#L597-L603)
3. [x] Actually use `ThreadedDownloadReporter`.
4. [x] Remove `VerboseDownloadReporter`.
5. [x] Verify that the [preexisting `req.spec.install` call](https://github.com/duckinator/rubygems/blob/db4a30ac3b5d901900e0f6777fc0b20bf5822fb0/lib/rubygems/request_set.rb#L195-L197) uses the cached version.
6. [x] When a gem is [being downloaded by the thread queue](https://github.com/duckinator/rubygems/blob/db4a30ac3b5d901900e0f6777fc0b20bf5822fb0/lib/rubygems/request_set.rb#L174), save it to the cache. (This is Kind Of Important&trade;.)
    * [x] Prerequisite: Figure out where the hell it's saving it currently?
7. [x] Determine if it handles `SIGINT`/Ctrl-C properly.

### CODE REVIEW

File: lib/rubygems/request_set.rb [hunk 1]
```diff
@@ -152,7 +152,34 @@ def install options, &block # :yields: request, installer
...
+        # means the thread should stop.
+        while req = download_queue.pop
+          break if req == :stop
+          req.spec.download options unless req.installed?
+        end
+      end
+    end
+
+    # Wait for all the downloads to finish before continuing
+    threads.map(&:value)
```
Comment (segiddins) [inline lib/rubygems/request_set.rb]: `.each` since we're not using the value

File: lib/rubygems/user_interaction.rb [hunk 2]
```diff
@@ -575,6 +575,57 @@ def done
...
+    def initialize(out_stream, *args)
+      @out = out_stream
+    end
+
+    ##
+    # Tells the download reporter that the +file_name+ is being fetched.
+    # The other arguments are ignored.
+
+    def fetch(file_name, *args)
+      puts "Fetching: #{file_name}"
```
Comment (segiddins) [inline lib/rubygems/user_interaction.rb]: should this be synchronized output?

File: lib/rubygems/user_interaction.rb [hunk 3]
```diff
@@ -575,6 +575,57 @@ def done
   end
 
   ##
+  # A progress reporter that behaves nicely with threaded downloading.
+
+  class ThreadedDownloadReporter
```
Comment (segiddins) [inline lib/rubygems/user_interaction.rb]: this class doesn't appear to be used anywhere

File: lib/rubygems/request_set.rb [hunk 4]
```diff
@@ -152,7 +152,34 @@ def install options, &block # :yields: request, installer
...
+      # is queued here so that there will be one symbol per thread.
+      download_queue << :stop
+
+      Thread.new do
+        # The pop method will block waiting for items, so the only way
+        # to stop a thread from running is to provide a final item that
+        # means the thread should stop.
+        while req = download_queue.pop
+          break if req == :stop
+          req.spec.download options unless req.installed?
```
Comment (segiddins) [inline lib/rubygems/request_set.rb]: will the spec be cached after being downloaded the first time?

File: lib/rubygems/user_interaction.rb (hunk 3 above)
Comment (duckinator) [inline lib/rubygems/user_interaction.rb]: Hm. I'm not sure how the default progress reporter will play with what we're doing; I suspect it'll look like a hot mess (hence why this class was created to start with). I also don't remember how the various `DownloadReporter` things tie into everything else. Will have to look into this in more detail.

File: lib/rubygems/request_set.rb (hunk 4 above)
Comment (duckinator) [inline lib/rubygems/request_set.rb]: I _think_ it defaults to downloading it into the cache, but I'll need to poke around to verify this.

**Comment by duckinator** [comment]
TODO:

* [ ] Verify that the spec stays in the cache after [being downloaded by the thread queue](https://github.com/duckinator/rubygems/blob/db4a30ac3b5d901900e0f6777fc0b20bf5822fb0/lib/rubygems/request_set.rb#L174).
* [ ] Verify that the [preexisting `req.spec.install` call](https://github.com/duckinator/rubygems/blob/db4a30ac3b5d901900e0f6777fc0b20bf5822fb0/lib/rubygems/request_set.rb#L195-L197) uses the cached version.
* [ ] Determine if/how `ThreadedDownloadReporter` should be used.
  * [ ] If it should be, actually do so. If not, remove it.
  * [ ] Use `locked_puts` in [`ThreadedDownloadReporter#fetch`](https://github.com/duckinator/rubygems/blob/db4a30ac3b5d901900e0f6777fc0b20bf5822fb0/lib/rubygems/user_interaction.rb#L597-L603)
* [ ] Specs!

**Comment by duckinator** [comment]
Rebased.

**Comment by duckinator** [comment]
It's theoretically possible to make `ThreadedDownloadReporter` report the download progress like `VerboseDownloadReporter` does, but that's so complicated that I think it should be a separate PR. I think the benefit of threaded downloads should (at least theoretically?) outweigh the cost of losing the progress indicators, but if someone disagrees let me know and we can try to figure out how to move forward with this.

As such, I'm going to switch to using ThreadedDownloadReporter, make a commit deleting VerboseDownloadReporter, then add a TODO saying to make ThreadedDownloadReporter report progress on systems that support it (with a link to the current implementation of VerboseDownloadReporter).

**Comment by indirect** [comment]
👍🏻

**Comment by duckinator** [comment]
Updated TODO:

* [ ] When a gem is [being downloaded by the thread queue](https://github.com/duckinator/rubygems/blob/db4a30ac3b5d901900e0f6777fc0b20bf5822fb0/lib/rubygems/request_set.rb#L174), save it to the cache. (This is Kind Of Important&trade;.)
* [ ] Verify that the [preexisting `req.spec.install` call](https://github.com/duckinator/rubygems/blob/db4a30ac3b5d901900e0f6777fc0b20bf5822fb0/lib/rubygems/request_set.rb#L195-L197) uses the cached version.
* [ ] Actually use `ThreadedDownloadReporter`.
    * [ ] Add TODO comment saying to re-add progress bars, with a link to the current implementation of VerboseDownloadReporter.
* [x] Use `locked_puts` in [`ThreadedDownloadReporter#fetch`](https://github.com/duckinator/rubygems/blob/db4a30ac3b5d901900e0f6777fc0b20bf5822fb0/lib/rubygems/user_interaction.rb#L597-L603)
* [ ] Tests!

**Comment by duckinator** [comment]
Rebased. I'm going to _try_ to get this wrapped up either today or Monday.

File: lib/rubygems/user_interaction.rb (hunk 3 above)
Comment (duckinator) [inline lib/rubygems/user_interaction.rb]: I explained how I'm planning to resolve this in [a comment on the PR itself](https://github.com/rubygems/rubygems/pull/1898#issuecomment-307607207). (Linking here so it's easier for people to find.)

**Comment by duckinator** [comment]
I looked into it a bit more, and here's all of the information I have about the implications of this PR on the `VerboseDownloadReporter`/nice progress indicators.

**TL;DR: I do not recommend re-adding the functionality of VerboseDownloadReporter after this PR is merged, due to complexity of making it work with concurrent downloads.**

### Overview

This PR adds the ability to download multiple gems concurrently. However, it will remove the nice progress indicators that the `VerboseDownloadReporter` has. [With the current implementation](https://github.com/rubygems/rubygems/blob/0a0ff34170eb1371068ff97a3911b484e9d812de/lib/rubygems/user_interaction.rb#L490-L531), those indicators will not work with concurrent downloads, and making them do so is a significant undertaking.

Accomplishing this requires rewriting lines that were previously printed, and tracking which file is referenced on which line.

I'd _optimistically_ expect this to require at least 200-300 lines of code, whereas VerboseDownloadReporter is currently a mere 74 lines.

Concurrent downloads themselves have already added enough, and I'm extremely weary of adding this much more code.

### Notes

<details>
<summary>Things to consider if this were to be implemented.</summary>
<br>

* Anything supporting ANSI escape codes is capable of this. However, you need to be careful to not use them when they're not supported, or you'll flood the users terminal with absolute garbage.
* You can not definitively determine if a program is supposed to actually support ANSI escape codes. You can, at best, go by what the `$TERM` environment variable claims. But this can be overridden trivially.
* In order to avoid getting garbled output due to race conditions, this will need to be managed by one thread that gets information from the download threads as it runs.
  * This thread would need to track the list of downloaded files in the order they're downloaded, and be able to update a line for an arbitrary file in that list.
  * It would also need to be able to accommodate new files being added to that list at any point.
* Given that RubyGems needs to avoid runtime dependencies, this will require adding a class or module that supports a subset of ANSI codes.
</details>

---

<details>
<summary>How I got that "200-300 lines of code" estimate.</summary>

<br>

The approach I considered was to whitelist `$TERM` environment variable values that indicate support of ANSI escape codes. For things that are in that whitelist, use a download reporter with thread-safe progress indicators. For things that are not in the whitelist, use the simpler reporter I've already added that just prints one message when it starts downloading, and one when it finishes.

You'd want a class, maybe something like `DownloadStatusReceiver`, that has mutex-locked methods that store messages about download status in a queue. It would have an array of references to files it has encountered, in the order download information was printed. It would then dequeue those messages and, for files it's encountered, update the relevant line; for files it has not encountered, add a reference to the aforementioned array and print the status.

(In the list of steps below, parenthesized text refers to items in this
table: https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_codes)

The process of updating a line would look something like this, surrounded by a mutex to avoid race conditions\*:

1. go up N lines (where N = distance from last line; use CSI Cursor Up),
2. move to column 1 (use CSI Cursor Horizontal Absolute),
3. clear from the cursor to the end of the (use CSI Erase In Line),
4. print the updated information,
5. move the cursor down N lines (same N as step 1; use CSI Cursor Down).

\* This mutex is why you don't want to print it in the download threads &mdash; if you put a mutex in those threads, you will wind up with one thread locking the mutex to print the results, and every other thread blocking while waiting on access to the mutex. My understanding is that this would happen every time any progress is made.

</details>

---

<details>
<summary>Information about `$TERM` and ANSI escape code support.</summary>

<br>

`$TERM` variable values that [indicate the terminal should support ANSI escape codes](https://en.wikipedia.org/wiki/Comparison_of_terminal_emulators#Capabilities ):

* ansi
* xterm
* vt100
* rxvt (or anything starting with "rxvt-" -- NOT mrxvt, however.)
 
There may be more, but that's a solid starting point.
</details>

**Comment by duckinator** [comment]
I'm having a lot of trouble grokking this part of codebase, and making significantly less progress than I expected because of it.

**Comment by indirect** [comment]
I'm totally okay with dropping accurate download progress indicators for each gem. Dramatically simpler options include:

1. printing undifferentiated dots from every download stream, until all downloads are done
2. printing a single message when the download starts, protected by a lock

I think either of those options would be fine. FWIW, option 2 would converge with the way Bundler prints download messages for threaded downloads.

**Comment by duckinator** [comment]
Rebased again.

**Comment by duckinator** [comment]
Picking this back up again. Gonna wrap up the stuff with `ThreadedDownloadReporter` first, then move to the rest of it.

**Comment by duckinator** [comment]
uhhh i'm not sure why that's failing. it's related to YAML safe loading.

**Comment by duckinator** [comment]
Alright. Well, in 4 hours I had 4 false starts, and ultimately accomplished nothing. I just cannot wrap my head around this part of the codebase. Didn't realize it'd been that long; should've probably stopped after like an hour.

This PR is pretty much stuck unless someone can help me with it.

I'm going to go ~launch myself into the sun~ take a break, now.

**Comment by duckinator** [comment]
Rebased. Still incomplete.

**Comment by duckinator** [comment]
Rebased. Gonna try doing some more work on this again.

**Comment by duckinator** [comment]
?! I THINK IT'S ACTUALLY WORKING?!

```
$ how_is
fish: Unknown command 'how_is'
$ ruby -Ilib bin/gem install how_is
Fetching faraday-0.12.2.gem
Fetching jwt-1.5.6.gem
Fetching public_suffix-3.0.3.gem
Fetching addressable-2.5.2.gem
Fetching multi_xml-0.6.0.gem
Fetching hashie-3.6.0.gem
Fetching multipart-post-2.0.0.gem
Fetching rack-2.0.5.gem
Fetching oauth2-1.4.0.gem
Fetching thread_safe-0.3.6.gem
Fetching descendants_tracker-0.0.4.gem
Fetching github_api-0.18.2.gem
Fetching contracts-0.16.0.gem
Fetching openssl-better_defaults-0.0.2.gem
Fetching cacert-0.5.0.gem
Fetching okay-7.0.0.gem
Fetching how_is-24.0.0.gem
Successfully installed public_suffix-3.0.3
Successfully installed addressable-2.5.2
Successfully installed hashie-3.6.0
Successfully installed multipart-post-2.0.0
Successfully installed faraday-0.12.2
Successfully installed jwt-1.5.6
Successfully installed multi_xml-0.6.0
Successfully installed rack-2.0.5
Successfully installed oauth2-1.4.0
Successfully installed thread_safe-0.3.6
Successfully installed descendants_tracker-0.0.4
Successfully installed github_api-0.18.2
Successfully installed contracts-0.16.0
Successfully installed openssl-better_defaults-0.0.2
Successfully installed cacert-0.5.0
Successfully installed okay-7.0.0
Successfully installed how_is-24.0.0
17 gems installed
$ how_is --version
24.0.0
$
```

!!! @indirect!

**Comment by duckinator** [comment]
And uninstalling then reinstalling it reuses the cache, as it should.

```
$ ruby -Ilib bin/gem install how_is
Successfully installed how_is-24.0.0
1 gem installed
$ ruby -Ilib bin/gem uninstall how_is
Remove executables:
	how_is

in addition to the gem? [Yn]  
Removing how_is
Successfully uninstalled how_is-24.0.0
$ ruby -Ilib bin/gem install how_is
Fetching how_is-24.0.0.gem
Successfully installed how_is-24.0.0
1 gem installed
$
```

It re-downloads the file itself, but not the dependency, which is exactly the same as the current behavior:

```
$ gem uninstall how_is
Remove executables:
	how_is

in addition to the gem? [Yn]  
Removing how_is
Successfully uninstalled how_is-24.0.0
$ gem install how_is
Fetching: how_is-24.0.0.gem (100%)
Successfully installed how_is-24.0.0
1 gem installed
$ 
```

File: lib/rubygems/config_file.rb [hunk 5]
```diff
@@ -43,6 +44,7 @@ class Gem::ConfigFile
   DEFAULT_BULK_THRESHOLD = 1000
   DEFAULT_VERBOSITY = true
   DEFAULT_UPDATE_SOURCES = true
+  DEFAULT_CONCURRENT_DOWNLOADS = 8
```
Comment (segiddins) [inline lib/rubygems/config_file.rb]: should this default to `nproc` instead of a fixed number?

File: lib/rubygems/request_set.rb [hunk 6]
```diff
@@ -152,7 +152,34 @@ def install options, &block # :yields: request, installer
...
+      download_queue << req
+    end
+
+    # Create N threads in a pool, have them download all the gems
+    threads = Gem.configuration.concurrent_downloads.times.map do
+      # When a thread pops this item, it knows to stop running. The symbol
+      # is queued here so that there will be one symbol per thread.
+      download_queue << :stop
+
+      Thread.new do
```
Comment (segiddins) [inline lib/rubygems/request_set.rb]: does this handle receiving `SIGINT` gracefully? (e.g. does the program exit as expected)

File: lib/rubygems/config_file.rb (hunk 5 above)
Comment (duckinator) [reply lib/rubygems/config_file.rb]: There's no reason to tie this to the number of processors/cores.

1. It's an I/O-bound task. I/O introduces a lot of waits, so even if you have a few more threads than the number of cores, it won't peg your CPU.
2. 16-core, 32-thread CPUs are a thing. `Etc.nprocessors` will return `32` on those systems. Using 32 threads for downloading files from a single server is a horrible, horrible idea.

There's also the fact that `Etc.nprocessors` doesn't work on Windows, so we'd have to roll our own thing for that.

File: lib/rubygems/request_set.rb (hunk 6 above)
Comment (duckinator) [reply lib/rubygems/request_set.rb]: I'm not sure. Added to the TODO list in the first comment.

Comment (duckinator) [reply lib/rubygems/request_set.rb]: @segiddins sorry for the delay &mdash; forgot to follow up on this. it does in fact handle `SIGINT` gracefully.

**Comment by indirect** [comment]
Nice! This is awesome, thanks for circling back around to finish things up.

@bundlerbot r+
